### PR TITLE
Get pseudobulk api

### PR DESCRIPTION
**Pull request by myushen**
solve issue #136 
get_pseudobulk API takes pre-calculated summarized experiment assays for selected sample IDs, combines them and returns a merged summarized experiment. The pre-calculated summarized experiment assays will be stored in a cloud container in the future.

### CODE REVIEW

File: R/counts.R [hunk 1]
```diff
@@ -180,6 +179,151 @@ get_single_cell_experiment <- function(
...
+#'   comparing a limited amount of genes, the "cpm" assay is more appropriate.
+#' @param repository A character vector of length one. If provided, it should be
+#'   an HTTP URL pointing to the location where the summarized experiment data is stored.
+#' @param cache_directory An optional character vector of length one. If
+#'   provided, it should indicate a local file path where any remotely accessed
+#'   files should be copied.
+#' @param features An optional character vector of features (ie genes) to return
+#'   the counts for. By default counts for all features will be returned.
+#' @returns A SummarizedExperiment object, with one assay for each value in the
+#'   assays argument
```
Comment (multimeric) [inline R/counts.R:203]: A lot of this is a copy from `get_single_cell_experiment`. I would consider reading https://cran.r-project.org/web/packages/roxygen2/vignettes/reuse.html. Specifically, I can see `@inheritParams` being useful.

File: R/counts.R [hunk 2]
```diff
@@ -180,6 +179,151 @@ get_single_cell_experiment <- function(
...
+#' @examples
+#' \dontrun{
+#' meta <- get_metadata() |> 
+#'         filter(sample_ == '068502277538ef5559154b543167fefa' | 
+#'                sample_ == '07605571f51519f71da03704f056fa43')
```
Comment (multimeric) [inline R/counts.R:208]:
```suggestion
#'         filter(sample_ %in% c('068502277538ef5559154b543167fefa', '07605571f51519f71da03704f056fa43'))
```

File: R/counts.R [hunk 3]
```diff
@@ -180,6 +179,151 @@ get_single_cell_experiment <- function(
...
+    has_name(raw_data, c("sample_"))
+  )
+  
+  versioned_cache_directory <- cache_directory
+  versioned_cache_directory |> dir.create(
+    showWarnings = FALSE,
+    recursive = TRUE
+  )
+  
+  subdirs <- assay_map[assays]
```
Comment (multimeric) [inline R/counts.R:263]: There is a lot of duplication of validation logic and other things here. I would rather see this refactored into a separate function that is called by both `get_single_cell_experiment` and `get_pseudobulk`.

**Review by multimeric — CHANGES_REQUESTED** [review]
Note also that `SingleCellExperiment` is a subclass of `SummarisedExperiment`, so if it's possible to write code that works on a `SummarisedExperiment`, then it will automatically work on both. This can help you simplify code and also documentation.

**Comment by stemangiola** [comment]
@multimeric any further concern?

**Comment by stemangiola** [comment]
I did not export fibrosis pseudobulk yet, @myushen did you test with fibrosis?

**Comment by myushen** [comment]
> I did not export fibrosis pseudobulk yet, @myushen did you test with fibrosis?

No, it is only based on cellxgene.

**Comment by multimeric** [comment]
I'm still concerned that `group_to_sme` is a near-exact copy of `group_to_sce` and the same for `get_pseudobulk` and `get_single_cell_experiment`. I see little reason why you can't consolidate these pairs of functions into one function, considering that you are working with a `SummarizedExperiment` in both cases.

**Comment by myushen** [comment]
> I'm still concerned that `group_to_sme` is a near-exact copy of `group_to_sce` and the same for `get_pseudobulk` and `get_single_cell_experiment`. I see little reason why you can't consolidate these pairs of functions into one function, considering that you are working with a `SummarizedExperiment` in both cases.

I'll wrap up into one function, but do we want to separate function names for get_sce and get_pseudobulk? @stemangiola

**Comment by stemangiola** [comment]
> > I'm still concerned that `group_to_sme` is a near-exact copy of `group_to_sce` and the same for `get_pseudobulk` and `get_single_cell_experiment`. I see little reason why you can't consolidate these pairs of functions into one function, considering that you are working with a `SummarizedExperiment` in both cases.
> 
> I'll wrap up into one function, but do we want to separate function names for get_sce and get_pseudobulk? @stemangiola

Are we asking here if

`get_single_cell_experiment` and `get_pseudobulk` should be the same function.

If this is the question I think that is beneficial keeping them separate.

File: R/counts.R [hunk 4]
```diff
@@ -9,115 +9,102 @@ NULL
...
+}
+
+#' Gets a Summarized Experiment from curated metadata
+#' 
+#' Given a data frame of Curated Atlas metadata obtained from [get_metadata()],
+#' returns a [`SummarizedExperiment::SummarizedExperiment-class`] object
+#' corresponding to the samples in that data frame
+#' @inheritParams param_validation
+#' @param type A character vector of length one. Either sce for Single Cell Experiment,
+#' or pseudobulk.
```
Comment (multimeric) [inline R/counts.R:79]:
```suggestion
#' @param type A character vector of length one. Either "sce" for Single Cell Experiment,
#' or "pseudobulk".
```

File: R/counts.R [hunk 5]
```diff
@@ -129,7 +116,7 @@ get_single_cell_experiment <- function(
     
     files_to_read <-
       raw_data |>
-      pull(.data$file_id_db) |>
+      pull(file_id_col) |>
```
Comment (multimeric) [inline R/counts.R:119]:
```suggestion
      pull(.data$file_id_col) |>
```

File: R/counts.R [hunk 6]
```diff
@@ -141,52 +128,57 @@ get_single_cell_experiment <- function(
...
+#' type. It requires specific columns for the data frame based on the type of
+#' experiment data being processed.
+#' 
 #' @param i Suffix to be added to the column names, to make them unique
 #' @param df The data frame to be converted
 #' @param dir_prefix The path to the single cell experiment, minus the final
 #'   segment
 #' @param features The list of genes/rows of interest
-#' @return A SingleCellExperiment object
+#' @return A Summarized Experiment object
```
Comment (multimeric) [inline R/counts.R:181]:
```suggestion
#' @return A `SummarizedExperiment` object
```

File: R/counts.R [hunk 7]
```diff
@@ -196,62 +188,91 @@ get_single_cell_experiment <- function(
...
+                         "cell_annotation_blueprint_singler",
+                         "cell_annotation_monaco_singler", 
+                         "cell_annotation_azimuth_l2",
+                         "cell_type_ontology_term_id",
+                         "sample_id_db")
+    
+    new_coldata <- df |> select(-dplyr::all_of(cell_level_anno)) |> distinct() |>
+      mutate(sample_identifier = glue("{sample_}___{cell_type_harmonised}"),
+             original_sample_id = .data$sample_identifier) |>
+      column_to_rownames("original_sample_id")
```
Comment (multimeric) [inline R/counts.R:262]: This is just for readability.
```suggestion
    new_coldata <- df |>
        select(-dplyr::all_of(cell_level_anno)) |>
        distinct() |>
        mutate(
            sample_identifier = glue("{sample_}___{cell_type_harmonised}"),
            original_sample_id = .data$sample_identifier
        ) |>
        column_to_rownames("original_sample_id")
```

**Review by multimeric — COMMENTED** [review]
Needs a test for the `get_pseudobulk` workflow.

Also this isn't a blocker, but I'd like to see some of the assay-specific logic (`if (type == "sce")`) moved into the calling function like `get_single_cell_experiment`, so that `get_summarized_experiment` remains decoupled from the subclass we're dealing with.

**Comment by stemangiola** [comment]
> Needs a test for the `get_pseudobulk` workflow.
> 
> Also this isn't a blocker, but I'd like to see some of the assay-specific logic (`if (type == "sce")`) moved into the calling function like `get_single_cell_experiment`, so that `get_summarized_experiment` remains decoupled from the subclass we're dealing with.

Thanks. Maybe @myushen do the tests, and we tackle the second point from @multimeric  in the future. We have to proceed, with other tasks.

**Review by multimeric — APPROVED** [review]

**Comment by stemangiola** [comment]
Great team work fellas!

**Pull request merged**